pelican/plugins/sitemap/sitemap.py

```diff
@@ -51,240 +47,177 @@ def format_date(date):
                 )
-                self.write_url(fake, fd)
-
-            # add template pages
-            # We use items for Py3k compat. .iteritems() otherwise
-            for path, template_page_url in self.context["TEMPLATE_PAGES"].items():
 
-                # don't add duplicate entry for index page
-                if template_page_url == "index.html":
-                    continue
-
-                fake = FakePage(
-                    status="published",
-                    date=self.now,
-                    url=template_page_url,
-                    save_as=template_page_url,
+                fd.write(
+                    XML_URL.format(
+                        siteurl,
+                        pageurl,
+                        lastmod,
+                        changefreq,
+                        priority,
+                        translations=translations,
+                    )
                 )
-                self.write_url(fake, fd)
 
-            for page in pages:
-                self.write_url(page, fd)
-
-            if self.format == "xml":
+            if is_xml:
                 fd.write(XML_FOOTER)
 
+        log.info("sitemap: Written {!r}".format(filename))
+
+    def _check_config(self, config):
+        if not isinstance(config, dict):
+            log.error("sitemap: The SITEMAP setting must be a dict")
+        for key in config.keys():
+            if key not in ("format", "exclude", "priorities", "changefreqs"):
+                log.error("sitemap: Invalid 'SITEMAP' key: {!r}".format(key))
+        changefreqs = config.get("changefreqs", {})
+        for key, value in changefreqs.items():
+            if key not in self.CHANGEFREQ_DEFAULTS:
+                log.error("sitemap: Invalid 'changefreqs' key: {!r}".format(key))
+            if value not in self.CHANGEFREQ_VALUES:
+                log.error("sitemap: Invalid 'changefreqs' value: {!r}".format(value))
+        for key, value in config.get("priorities", {}).items():
+            if key not in self.PRIORITY_DEFAULTS:
+                log.error("sitemap: Invalid 'priorities' key: {!r}".format(key))
+        fmt = config.get("format")
+        if fmt not in (None, "txt", "xml"):
+            log.error(
+                "sitemap: Invalid 'format' value: {!r}; ",
+                "must be 'txt' or 'xml'".format(fmt),
+            )
+        exclude = config.get("exclude", ())
+        if not all(isinstance(i, str) for i in exclude):
+            log.error(
+                "sitemap: Invalid 'exclude' value: {!r}; ",
+                "must be a list of str".format(exclude),
+            )
+
 
-def get_generators(generators):
-    return SitemapGenerator
+generator = SitemapGenerator()
 
 
 def register():
-    signals.get_generators.connect(get_generators)
+    signals.initialized.connect(generator.init)
+    signals.content_written.connect(generator.queue_page)
+    signals.finalized.connect(generator.finalize)
```

Comment:
Signal `finalized` doesn't appear to be triggered when running `pelican --autoreload`. Is `article_writer_finalized` signal a suitable alternative?

Reply:
Are you sure? It should always be triggered.

Reply:
Somewhat sure. `pelican -rD`: breakpoint in `SitemapGenerator.finalized()` is never hit.
```
$ pelican -rD
DEBUG: Pelican version: 4.2.0
...
-> Copying ... to ...
Done: Processed 4 articles, 0 drafts, 10 pages, 1 hidden page and 0 draft pages in 0.21 seconds.
```
```bash
$ cat output/sitemap.txt
output/sitemap.txt: No such file or directory
```

Reply:
I need a bit more to go on:

```console
$ cat pelicanconf.py
from pelican import signals
import logging

logger = logging.getLogger(__name__)

class FinalizedTest:
    __name__ = 'FinalizedTestPlugin'
    def finalized(self, pelican):
        logger.warning('Finalized reached')

    def register(self):
        signals.finalized.connect(self.finalized)


PLUGINS = [FinalizedTest()]
$ pelican content -r
WARNING: Feeds generated without SITEURL set properly may not be valid
WARNING: No timezone information specified in the settings. Assuming your timezone is UTC for feed generation. Check https://docs.getpelican.com/en/latest/settings.html#TIMEZONE for more information
  --- AutoReload Mode: Monitoring `content`, `theme` and `settings` for changes. ---

-> Modified: settings, content, theme. re-generating...
WARNING: Finalized reached
Done: Processed 1 article, 0 drafts, 0 pages, 0 hidden pages and 0 draft pages in 0.11 seconds.
```

Reply:
Ah, my bad. The new sitemap build machinery based on poetry doesn't allow `--editable` installs, so my source changes weren't picked up immediately without a rebuild. Here's a related [position](https://github.com/python-poetry/poetry/issues/34#issuecomment-379361055) of a poetry leading dev. God help us.

Reply:
So this was made to accommodate _i18n_subsites_, which spawns a complete separate build process for each language. I'm no longer using i18n_subsites as the plugin proved detrimental in other places (with other plugins, namely _process_images_), so I'm happy to revert this messy part.

That said, I managed to replicate the i18n subsites behavior with existing pelican core configuration options alone. I think the core might actually benefit from the very little extra needed to elegantly sustain multi-language sites without necessitating i18n_subsites or hacking-sewing together several distinct site builds. I think what's missing in core mostly boils down to a new `LANG_OVERRIDES` setting exactly akin to [`I18N_SUBSITES` dict](https://github.com/getpelican/pelican-plugins/tree/master/i18n_subsites#setting-it-up) and some template changes. Before seeing any code to that effect, what do you think, @justinmayer, @avaris ?

Reply:
Regarding code above, you probably need to reset `bool` values once you process. 

```python
    def finalize(self, pelican):
        self.pelican_finalized[pelican] = True
        # Wait for all i18n_subsites to finish
        if all(self.pelican_finalized.values()):
            self._write_out()
            for p in self.pelican_finalized:
                sef.pelican_finalized[p] = False
```

Reply:
I don't think so. That dict only serves as a counting semaphore. If anything, after `finalized` I could just `self.pelican_finalized.clear()`. But the inner `if all` isn't ever even hit (with `--autoreload`, which this thread is about).

Reply:
OK, I played with it a bit more. I think there is some "issue" in `autoreload`. Every time a pelican instance is created (twice initially and one more for each time settings changes), original connection persists and signals are connected again. I'm not sure why previous ones persist. I would expect them to be garbage collected. I need to look into it more.

```console
$ cat pelicanconf.py
from pelican import signals
import logging
import uuid

logger = logging.getLogger(__name__)

class FinalizedTest:
    __name__ = 'FinalizedTestPlugin'
    def finalized(self, pelican):
        logger.warning('Finalized reached: %s %s', id(self), uuid.uuid4())

    def register(self):
        signals.finalized.connect(self.finalized)


PLUGINS = [FinalizedTest()]
$ pelican content -r
WARNING: Feeds generated without SITEURL set properly may not be valid
WARNING: No timezone information specified in the settings. Assuming your timezone is UTC for feed generation. Check https://docs.getpelican.com/en/latest/settings.html#TIMEZONE for more information
  --- AutoReload Mode: Monitoring `content`, `theme` and `settings` for changes. ---

-> Modified: settings, content, theme. re-generating...
WARNING: Finalized reached: 140251169452240 2db9b27b-bcb6-4344-b207-63d596c7ab1b
WARNING: Finalized reached: 140251169441776 d585796c-3734-4bf0-a7b6-23b31f375515
Done: Processed 2 articles, 0 drafts, 0 pages, 0 hidden pages and 0 draft pages in 0.13 seconds.

-> Modified: settings. re-generating...
WARNING: Finalized reached: 140251169452240 ce30b9ab-6c34-4e7a-8083-b39b0051f246
WARNING: Finalized reached: 140251169440240 3c21e73e-e167-4342-829a-46a62eb3c4e8
WARNING: Finalized reached: 140251169441776 f59128a9-97e1-454e-a4d4-81c301ac4f82
Done: Processed 2 articles, 0 drafts, 0 pages, 0 hidden pages and 0 draft pages in 0.12 seconds.

-> Modified: settings. re-generating...
WARNING: Finalized reached: 140251169452240 0915f645-c34a-4bcc-be3e-06be9ae14e58
WARNING: Finalized reached: 140251169440240 9aff32ce-1d10-4332-83af-f152e601a19a
WARNING: Finalized reached: 140251167897920 325cea6a-76c2-48a4-992d-75ef5213a986
WARNING: Finalized reached: 140251169441776 c2693eaf-23e4-4117-8ecd-6e1f496c7d3a
Done: Processed 2 articles, 0 drafts, 0 pages, 0 hidden pages and 0 draft pages in 0.09 seconds.
```

Reply:
@kernc Well, my experiment above suggests it is called too much instead :). I'd suggest doing tests *without* `i18n_subsites` first. That introduces way too much complexity to debug.

Reply:
> The new sitemap build machinery based on poetry doesn't allow `--editable` installs, so my source changes weren't picked up immediately without a rebuild. Here's a related [position](https://github.com/python-poetry/poetry/issues/34#issuecomment-379361055) of a poetry leading dev. God help us.

I hear you. Although I don't have many peeves with Poetry, this is probably my biggest one. It is not at all clear to me why Sébastien steadfastly refuses to support editable installation of arbitrary packages. In this case, the primary workaround is to `poetry install` the plugin, uninstall Pelican, and re-install Pelican in editable mode. Other longer-term options include changing pyproject.toml dependencies to refer to relative folder paths, generating and including setup.py files with all plugins, or ditching Poetry entirely. I'm not really jazzed about any of these latter longer-term options, so for now I am content to use the primary workaround until (hopefully) a better option presents itself.

Reply:
I think the issue here is that there are `pelican` instances created (i.e. `initialized` is fired) that do not fire any `finalized` signals. There is at least one too many `pelican` instance creation that could be easily eliminated (reduce from three to two). But this would not solve the underlying issue. So, yes, your `all` would never be `True` in this case. I still need to investigate further but I think `signal`s should be reset (recreate or disconnect everything) when a `pelican` instance is created.

Reply:
The above is good and all but it won't solve the issue here :). You'll still "store" `pelican` instances that won't do anything. I think the solution here would still entail some state reset. `register` is probably the place to do that.

Reply:
How would this work for _i18n_subsites_, which issues `register` for each
subsite, but we'd prefer to generate just one, conjoined sitemap at the end?

Reply:
I'm looking at `i18n_subsites` code (always gives me a mild headache :)), and *I think* it goes around this by hooking up the initializer to `get_generators` instead of `initialized`. `get_generators` is fired at the start of `run` so it is guaranteed to emit `finalized` as well.

Moreover, based on the code, "main" pelican will always be the first to start and last to end. Combining all of these, *I think* the following will work:
```python

class SitemapGenerator:
    def __init__(self):
        # ...
        self._main_pelican = None

    def init(self, pelican):
        # ...
        if self._main_pelican is None:
            self._main_pelican = pelican

    def finalize(self, pelican):
        if pelican == self._main_pelican:
            self._write_out()
            self._main_pelican = None  # reset for autoreload

    # ...

def register():
    signals.get_generators.connect(generator.init)
    # ...
    signals.finalized.connect(generator.finalize)
```

Reply:
Moreover, I may also figured out why previous connections were persisting. It is partially because of my (ab)use of "`class`-as-plugin". Although, that's still a semi-valid way to define plugin so I'd rather keep it working :). But that's a fix for `pelican` itself, not related here. module/package style plugins are fine since they are cached by python and calling `connect` multiple times with same function doesn't register multiple copies. But my class-like plugins were "fresh" each time since they are instances created on the fly.

Reply:
@avaris Looking into it now (:sweat_smile:), I don't see how:

> "main" pelican will always be the first to start

The order of generation seems to be determined (assuming Py3.6+) by entry definition order in `I18N_SUBSITES`? :thinking: :confused: 

https://github.com/getpelican/pelican-plugins/blob/4181b14b0ffdac42285d5940a94c0ce4dc8ac1b0/i18n_subsites/i18n_subsites.py#L428
https://github.com/getpelican/pelican-plugins/blob/4181b14b0ffdac42285d5940a94c0ce4dc8ac1b0/i18n_subsites/i18n_subsites.py#L71

Reply:
@kernc yes, that is true for the subsites. Whole thing works like a recursive function, triggered by [`get_writers`](https://github.com/getpelican/pelican-plugins/blob/4181b14b0ffdac42285d5940a94c0ce4dc8ac1b0/i18n_subsites/i18n_subsites.py#L444) signal. It starts with the "main" site, that is the `pelican` instance responsible with the original settings. Once it reaches "write" state (`get_writers`), it spawns another pelican instance using one of the subsites. Once that reaches `get_writers`, spawns another subsite... so on, until it exhausts all subsites. At that point it needs to update intrasite links and continue writing. So flow is like this roughly, (assuming 3 subsites):

```
__ main site (up to get_writers)
    \_ subsite1 (up to get_writers)
        \_ subsite2 (up to get_writers)
            \_ subsite3 (up to get_writers)
                \_ update links
            \_ write and finalize subsite3
        \_ write and finalize subsite2
    \_ write and finalize subsite1
\_ write and finalize main site
```

Order of main is always "start first, end last", it's basically the entry and exit point for recursion. But, yes, order of subsites in the middle is undefined, i.e. subject to python dict restrictions.

Reply:
Thanks. I'll take your word for it. :grin:

I might be encouraged to write-up a three point issue on how Pelican can support language subsites without this antagonistic plugin.